contract.sol

```diff
@@ -0,0 +1,152 @@
+            amount -= penalty;  // Subtract penalty from the amount to be withdrawn
+            accumulatedPenalties += penalty;  // Add penalty to accumulated penalties
+        }
+
+        userStake.amount -= amount;  // Update the user's staked amount
+        stakingToken.transfer(msg.sender, amount);  // Transfer the remaining amount to the user
+
+        emit Withdrawn(msg.sender, amount, penalty, rewardToken);  // Emit a Withdrawn event
+    }
+
+    // Function to claim rewards based on staked tokens
+    function claimReward(address rewardToken) external nonReentrant whenNotPaused {
+        Stake storage userStake = stakes[msg.sender][rewardToken];  // Access the user's stake for this reward token
+        if (userStake.amount == 0) revert NoStakeFound();  // Ensure the user has a stake
+        if (block.timestamp <= userStake.startTime + minStakingPeriod) revert CannotClaimRewardYet();  // Ensure the minimum staking period has passed
+
+        uint256 stakingDuration = block.timestamp - userStake.startTime;  // Calculate the staking duration
+        uint256 rewardAmount = (userStake.amount * rewardRate * stakingDuration) / 1e18;  // Calculate the reward amount
+
+        IERC20(rewardToken).transfer(msg.sender, rewardAmount);  // Transfer the reward tokens to the user
+
+        userStake.startTime = block.timestamp;  // Reset the stake start time
+
+        emit RewardClaimed(msg.sender, rewardAmount, rewardToken);  // Emit a RewardClaimed event
+    }
+
+    // Function for the contract owner to withdraw accumulated penalties
+    function withdrawPenalties() external onlyOwner nonReentrant whenNotPaused {
+        uint256 amountToWithdraw = accumulatedPenalties;  // Get the total accumulated penalties
+        if (amountToWithdraw == 0) revert NoPenaltiesToWithdraw();  // Ensure there are penalties to withdraw
+
+        accumulatedPenalties = 0;  // Reset the accumulated penalties
+        stakingToken.transfer(owner(), amountToWithdraw);  // Transfer the penalties to the owner
+
+        emit PenaltiesWithdrawn(owner(), amountToWithdraw);  // Emit a PenaltiesWithdrawn event
+    }
+
+    // Function to add a new reward token
+    function addRewardToken(address rewardToken) external onlyOwner {
+        if (isRewardToken[rewardToken]) revert InvalidRewardToken();  // Ensure the token is not already a reward token
+        rewardTokens.push(rewardToken);  // Add the token to the reward tokens list
+        isRewardToken[rewardToken] = true;  // Mark the token as a valid reward token
+    }
+
+    // Function to set the minimum staking period
+    function setMinStakingPeriod(uint256 period) external onlyOwner {
+        minStakingPeriod = period;
+    }
+
+    // Function to set the early withdrawal penalty percentage
+    function setEarlyWithdrawalPenalty(uint256 penalty) external onlyOwner {
+        earlyWithdrawalPenalty = penalty * 1e18;
```

Comment:
You can simply keep the penalty as 1, 10 ,100, 1000, 10000 like this , no need to scale so much.
And for DENOMINATOR you can use 10000, we don't want to go beyond 1/10000 decimals.

Reply:
this is also not done